src/net_processing.cpp

```diff
@@ -1327,7 +1326,7 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     }
 
     m_connman.ForEachNode([this, &pcmpctblock, pindex, &msgMaker, fWitnessEnabled, &hashBlock](CNode* pnode) {
-        AssertLockHeld(cs_main);
+        LockAssertion lock(::cs_main);
```

Comment:
What's the reason to prefer LockAssertion to AssertLockHeld if AssertLockHeld works at compile time and runtime and LockAssertion only works at compile time? Maybe this could be clarified https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization

Reply:
When TSA fails to detect that a mutex is locked we do want to use `LockAssertion` as using `AssertLockHeld` no longer helps:
```
$ git diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 0e049bd66..90bd626fc 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -1370,7 +1370,7 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     }
 
     m_connman.ForEachNode([this, &pcmpctblock, pindex, &msgMaker, fWitnessEnabled, &hashBlock](CNode* pnode) {
-        LockAssertion lock(::cs_main);
+        AssertLockHeld(::cs_main);
 
         // TODO: Avoid the repeated-serialization here
         if (pnode->nVersion < INVALID_CB_NO_BAN_VERSION || pnode->fDisconnect)
$ make
net_processing.cpp:1373:9: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::recursive_mutex> >' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        AssertLockHeld(::cs_main);
        ^
./sync.h:79:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
net_processing.cpp:1378:9: warning: calling function 'ProcessBlockAvailability' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        ProcessBlockAvailability(pnode->GetId());
        ^
net_processing.cpp:1379:30: warning: calling function 'State' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        CNodeState &state = *State(pnode->GetId());
                             ^
net_processing.cpp:1383:18: warning: calling function 'PeerHasHeader' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
                 ^
net_processing.cpp:1383:51: warning: calling function 'PeerHasHeader' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
                                                  ^
5 warnings generated.
```

Reply:
And this `LockAssertion` application is documented: https://github.com/bitcoin/bitcoin/blob/a0a422c34cfd6514d0cc445bd784d3ee1a2d1749/src/sync.h#L355-L357

Reply:
> Maybe this could be clarified https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization

Mind suggesting the clarification as my English is so poor?

Reply:
Oh, I wasn't aware `AssertLockHeld` annotations changed from `ASSERT_EXCLUSIVE_LOCK` 
to `EXCLUSIVE_LOCKS_REQUIRED` in #19668. This seems like a questionable decision to me, because now instead of being able to use `AssertLockHeld` everywhere, we have separate confusing `AssertLockHeld` and `LockAssertion` calls.

I think I would want to clarify this by changing `AssertLockHeld` annotations back to `ASSERT_EXCLUSIVE_LOCK`, making it a weaker assertion that only causes errors at runtime, never compile time. Then we could drop LockAssertion and s/LockAssertion/AssertLockHeld/ and only have one type of assertion to think about.

If this doesn't work because it interferes with work you are doing, another option 
might be to s/LockAssertion/WeaklyAssertLockHeld/ everywhere. (Also maybe simplify the implementation to be an `ASSERT_EXCLUSIVE_LOCK` function instead of a class instance that requires a variable declaration.) The new naming should make it clearer which call is preferred:

* `AssertLockHeld` - stricter assertion which causes errors at compile time if the compiler supports annotations, or causes errors at runtime if the compiler doesn't support annotations
* `WeaklyAssertLockHeld` - less strict alternative which only triggers errors at runtime, never compile-time, and can be used when it's not possible for compiler to determine that the mutex is locked.

Reply:
We have three types of assertions related for locking:

* marking a function as `EXCLUSIVE_LOCKS_REQUIRED` has a compile time check with clang that the caller has already obtained the lock
* declaring a `LockAssertion` instance overrides the clang compile time checks by saying that we've already somehow acquired the lock in a way that we can't prove via clang thread safety annotations. This is useful for lambda functions (which could be annotated) that are called from some generic dispatcher like `ForEach` (that can't be annotated).
* adding a call to `AssertLockHeld` does a runtime check if DEBUG_LOCKORDER is enabled. it's annotated with EXCLUSIVE_LOCKS_REQUIRED so is redundant with the compile time checks, but the compile time checks aren't available outside of clang. it's automatically called from LockAssertion to ensure that the compile time checks aren't overridden incorrectly.

As a result:

* using AssertLockHeld everywhere remains fine, and no longer needlessly disables the compile time checks
* LockAssertion should only be used very rarely -- and in fact it's only used in net_processing in some ForEach/ForEachNode functions.

Maybe we should move the locking outside of the ForEach functions into the caller, and convert them into `for ( x : nodes ) { ... }` instead, in which case we could probably do away with LockAssertion entirely.

Reply:
> no longer needlessly disables the compile time checks

It's funny the only reason I ever called it was to disable compile time checks, and now it doesn't do that anymore! :smile: If really the only thing AssertLockHeld is doing is adding runtime checks on top of compile time checks that are already present, I think that is nearly useless. Anyway, can take it up in another PR.

Reply:
> Anyway, can take it up in another PR.

Created #19865 to experiment

Reply:
You use `LockAssertion g(cs_foo);` to tell the compiler that a lock is guaranteed to already be held within a code block even though the function annotations don't guarantee it. AssertLockHeld is useful if you're not compiling with clang.

Reply:
> AssertLockHeld is useful if you're not compiling with clang.

Yes, that's what I mean by "nearly useless." It can help developers who aren't using clang, and who are compiling with DEBUG_LOCKORDER, and who want to wait for an assert to trigger at runtime. But it won't make the codebase more thread safe because we already don't merge changes that don't compile with clang.

Any case, if you see problems with #19865, it would be good to comment there. I'm not an expert on lock annotations, so maybe the AssertLockHeld calls removed in #19865 are actually doing something useful besides haphazardly repeating compile time assertions at runtime.

Reply:
> > AssertLockHeld is useful if you're not compiling with clang.
> 
> Yes, that's what I mean by "nearly useless." It can help developers who aren't using clang, and who are compiling with DEBUG_LOCKORDER, and who want to wait for an assert to trigger at runtime. But it won't make the codebase more thread safe because we already don't merge changes that don't compile with clang.

I [disagree](https://github.com/bitcoin/bitcoin/pull/19865#issuecomment-686986462) with statements about "useless" `AssertLockHeld`.